components/omega/src/base/MachEnv.cpp

```diff
@@ -0,0 +1,441 @@
+   MPI_Comm_dup(InComm, &(DefEnv.Comm));
+
+   // get task ID and set local MPI task
+   MPI_Comm_rank(DefEnv.Comm, &(DefEnv.MyTask));
+
+   // get total number of MPI tasks
+   MPI_Comm_size(DefEnv.Comm, &(DefEnv.NumTasks));
+
+   // Set task 0 as master
+   DefEnv.MasterTask = 0;
+
+   // determine if this task is the master task
+   if (DefEnv.MyTask == DefEnv.MasterTask) {
+      DefEnv.MasterTaskFlag = true;
+   } else {
+      DefEnv.MasterTaskFlag = false;
+   }
+
+   // All tasks are members of this communicator's group
+   DefEnv.MemberFlag = true;
+
+#ifdef OMEGA_THREADED
+   // total number of OpenMP threads
+   DefEnv.NumThreads = omp_get_num_threads();
+#else
+   DefEnv.NumThreads = 1;
+#endif
+
+   // Add to list of environments
+   OMEGA::MachEnv::addEnv("Default", &DefEnv);
+
+   // Retrieve this environment and set pointer to DefaultEnv
+   MachEnv::DefaultEnv = getEnv("Default");
+
+} // end init Mach Env
+
+// Create functions
+//------------------------------------------------------------------------------
+// Create a new environment from a contiguous subset of an
+// existing environment starting from same root task.
+
+void MachEnv::createEnv(const std::string Name, // [in] name of environment
+                        const MachEnv *InEnv,   // [in] parent MachEnv
+                        const int NewSize       // [in] num tasks in new env
+) {
+
+   // Create the new Env structure and make it persistent (static)
+   static MachEnv NewEnv;
+
+   // Get communicator from old environment
+   MPI_Comm InComm = InEnv->getComm();
+
+   // Error checks on new size
+   int OldSize;
+   MPI_Comm_size(InComm, &OldSize);
+   if (NewSize > OldSize) {
+      std::cerr << "Invalid NewSize in MachEnv constructor "
+                << "NewSize = " << NewSize << " is larger than old size"
+                << OldSize << std::endl;
+      return;
+   }
+
+   // First retrieve the group associated with the old environment
+   MPI_Group InGroup;
+   MPI_Comm_group(InComm, &InGroup);
+
+   // Define the range of tasks in new group (0, NewSize-1)
+   int NRanges     = 1;
+   int LastTask    = NewSize - 1;
+   int Range[1][3] = {0, LastTask, 1};
+
+   // Create a new group with the new range
+   MPI_Group NewGroup;
+   MPI_Group_range_incl(InGroup, NRanges, Range, &NewGroup);
+
+   // Create the communicator for the new group
+   MPI_Comm_create(InComm, NewGroup, &NewEnv.Comm);
+
+   // determine whether this task is a part of the new communicator
+   if (NewEnv.Comm != MPI_COMM_NULL) {
+
+      NewEnv.MemberFlag = true;
+      // get task ID for local MPI task
+      MPI_Comm_rank(NewEnv.Comm, &NewEnv.MyTask);
+      // get total number of MPI tasks
+      MPI_Comm_size(NewEnv.Comm, &NewEnv.NumTasks);
+
+   } else {
+
+      NewEnv.MemberFlag = false;
+      NewEnv.MyTask     = -999;
+      NewEnv.NumTasks   = -999;
+   }
+
+   // Set task 0 as master
+   NewEnv.MasterTask = 0;
```

Comment:
Just wondering if we should make master rank an optional argument with a default value of zero?

Reply:
Sure, I will add.